src/block/Beacon.php

```diff
@@ -23,11 +23,176 @@
+		BlockLegacyIds::IRON_BLOCK => true,
+		BlockLegacyIds::GOLD_BLOCK => true,
+		BlockLegacyIds::DIAMOND_BLOCK => true,
+		BlockLegacyIds::EMERALD_BLOCK => true
+		//TODO netherite block
+	];
+
+	private const ALLOWED_PRIMARY_EFFECTS = [
+		EffectIds::HASTE => true,
+		EffectIds::JUMP_BOOST => true,
+		EffectIds::RESISTANCE => true,
+		EffectIds::SPEED => true,
+		EffectIds::STRENGTH => true
+	];
+
+	private const ALLOWED_SECONDARY_EFFECTS = [
+		EffectIds::REGENERATION => true
+	];
+
+	private int $primaryEffect;
+	private int $secondaryEffect;
+
+	public function readStateFromWorld() : void{
+		parent::readStateFromWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			$this->primaryEffect = $tile->getPrimaryEffect();
+			$this->secondaryEffect = $tile->getSecondaryEffect();
+		}
+	}
+
+	public function writeStateToWorld() : void{
+		parent::writeStateToWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			$tile->setPrimaryEffect($this->primaryEffect);
+			$tile->setSecondaryEffect($this->secondaryEffect);
+		}
+	}
+
+	public function getPrimaryEffect() : int{
+		return $this->primaryEffect;
+	}
+
+	/** @return $this */
+	public function setPrimaryEffect(int $primaryEffect) : self{
+		if(!isset(self::ALLOWED_PRIMARY_EFFECTS[$primaryEffect])){
+			throw new \InvalidArgumentException("Effect ID \"$primaryEffect\" is not allowed in the primary effect");
+		}
+		$this->primaryEffect = $primaryEffect;
+		return $this;
+	}
+
+	public function getSecondaryEffect() : int{
+		return $this->secondaryEffect;
+	}
+
+	/** @return $this */
+	public function setSecondaryEffect(int $secondaryEffect) : self{
+		if(!isset(self::ALLOWED_PRIMARY_EFFECTS[$secondaryEffect]) || !isset(self::ALLOWED_SECONDARY_EFFECTS[$secondaryEffect])){
+			throw new \InvalidArgumentException("Effect ID \"$secondaryEffect\" is not allowed in the secondary effect");
+		}
+		$this->secondaryEffect = $secondaryEffect;
+		return $this;
+	}
 
 	public function getLightLevel() : int{
 		return 15;
 	}
 
-	//TODO
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if($player instanceof Player){
+			$player->setCurrentWindow(new BeaconInventory($this->position));
+		}
+
+		return true;
+	}
+
+	public function onScheduledUpdate() : void{
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 20 * 3);
+
+		if(!$this->viewSky()){
+			return;
+		}
+
+		$effectIdMap = EffectIdMap::getInstance();
+		$primaryE = $effectIdMap->fromId($this->primaryEffect);
+		$secondaryE = $effectIdMap->fromId($this->secondaryEffect);
+		if($primaryE === null && $secondaryE === null){
+			return;
+		}
+
+		$beaconLevel = 0;
+		for($i = 1; $i <= 4; $i++){
+			if(!$this->isBeaconLevelValid($i)){
+				break;
+			}
+			$beaconLevel++;
+		}
+		if($beaconLevel > 0){
+			$radius = (10 * $beaconLevel) + 10;
```

Comment:
This could just be 11 x radius.

Reply:
If we put just 11x radius, for level 1, it will be 11 of range but 20 needed

Reply:
Defining the variable `$beaconLevel` as: `$beaconLevel = 1` could simplify this expression to: `$radius = 10 * $beaconLevel;`. Also don’t forget to also check `if($beaconLevel > 1)`

Reply:
My bad. My brain isn't working very well right now.